Feat: Implement the simple statistical member functions

ThreeMonth03 opened the pull request:
According to issue #513, I implement the naive function without considering the array arguments.

ThreeMonth03 commented on cpp/modmesh/buffer/SimpleArray.hpp:
```diff
@@ -123,6 +123,18 @@ class SimpleArrayMixinModifiers
+{
+    using type = U;
+};
+
```
Determine whether the type is complex by SFINAE.

yungyuc replied:
I propose to name it as `select_real_t`.

ThreeMonth03 commented on cpp/modmesh/buffer/SimpleArray.hpp:
```diff
@@ -193,6 +207,85 @@ class SimpleArrayMixinCalculators
+            else
+            {
+                acc += athis->data(i) * athis->data(i);
+            }
```
Var(complex_number) = Var(complex_number.real) + Var(complex_number.imag)

yungyuc replied:
> Var(complex_number) = Var(complex_number.real) + Var(complex_number.imag)

I am not so sure and need a proof.  @Gene0315 , could you please help find a solid reference for the definition of complex number variance.

A [webpage I got online](https://dsp.stackexchange.com/questions/40306/what-would-be-the-variance-for-complex-number) says

$$
\sum_{n=1}^N x_nx_n^H = \frac{N}{2}\sigma_x^2
$$

where $\sigma_x^2$ is the variance.  $\square^H$ is the complex conjugate.

@j8xixo12 @KHLee529 FYI.

ThreeMonth03 commented on cpp/modmesh/math/Complex.hpp:
```diff
@@ -44,6 +44,20 @@ struct ComplexImpl
+        : real_v(t)
+        , imag_v(0.0)
+    {
+    }
```
In order to calculate the `SimpleArray::mean()`, I implement the constructor of `ComplexImpl` with one argument.

ThreeMonth03 commented on tests/test_buffer.py:
```diff
@@ -970,6 +970,96 @@ def test_abs(self):
+        npmed = np.median(nparr)
+        smed = sarr.median()
+        self.assertEqual(npmed.real, smed.real)
+        self.assertEqual(npmed.imag, smed.imag)        
```
Check `median` of complex array.

ThreeMonth03 commented on tests/test_buffer.py:
```diff
@@ -970,6 +970,96 @@ def test_abs(self):
+        sarr = modmesh.SimpleArrayComplex128(array=nparr)
+        npvar = np.var(nparr)
+        svar = sarr.var()
+        self.assertEqual(npvar, svar)
```
Check the variance of complex array.

ThreeMonth03 reviewed (commented):
@yungyuc You can review this pull request if the CI pass. Thank you.

yungyuc commented on cpp/modmesh/buffer/SimpleArray.hpp:
```diff
@@ -193,6 +207,85 @@ class SimpleArrayMixinCalculators
+    value_type median() const
+    {
+        auto athis = static_cast<A const *>(this);
+        A acopy(*athis);
```
Don't make copy.  It's super slow and impossible to be sped up using SIMD.  Use in-place operations.  Do not rely on STL.

ThreeMonth03 replied:
I remember that `std::nth_element()` would rearrange the data. I think it is impossible to implement in-place `median()` without modifying the buffer.

yungyuc replied:
Don't use STL.  Write pointer code.  Even if you need a new buffer, I do not want STL code which is highly unlikely to be able to be migrated to SIMD and accelerators.

ThreeMonth03 commented on cpp/modmesh/math/Complex.hpp:
```diff
@@ -121,13 +135,21 @@ struct ComplexImpl
+    if (lhs.real_v > rhs.real_v)
+        return false;
+    return lhs.imag_v <= rhs.imag_v;
 }
```
In numpy, the default relational operator is compared by lexicographical order. No matter `np.min()` or `np.median()`, they are obeyed with the lexicographical order by default.

yungyuc commented on cpp/modmesh/math/Complex.hpp:
```diff
@@ -44,6 +44,20 @@ struct ComplexImpl
+    {
+    }
+
+    ComplexImpl(const T & t)
```
`T` is a fundamental type and don't use a reference.

ThreeMonth03 replied:
I'm curious about when to use the constant reference for the parameter.

yungyuc replied:
When it is not a fundamental type.

ThreeMonth03 reviewed (commented):
@yungyuc You can review this pull request if the CI pass. Thank you.

yungyuc commented on cpp/modmesh/math/Complex.hpp:
```diff
@@ -121,13 +135,21 @@ struct ComplexImpl
 bool operator<(const ComplexImpl<T> & lhs, const ComplexImpl<T> & rhs)
 {
-    return lhs.norm() < rhs.norm();
+    if (lhs.real_v < rhs.real_v)
```
Always use curly braces `{}` for a single-line block.

Comparison of complex numbers are trick.  Perhaps we should not provide the greater than and the less than operators.  We'll see.  Please leave a comment to document the trickiness.

yungyuc commented on tests/test_buffer.py:
```diff
@@ -970,6 +970,106 @@ def test_abs(self):
+        self.assertEqual(npmed.real, smed.real)
+        self.assertEqual(npmed.imag, smed.imag)
+
+    def test_average(self):
```
Include tests for 1D arrays for all the operations.

yungyuc reviewed (changes requested):
Some points to address:

- [x] Rename `real_type_selector` to `select_real_t`.  Or we can discuss what is a better name for the template.
- [x] Use in-place operation as much as possible.  It means you usually need to write pointer code.
- [x] Don't use reference for a fundamental type argument.
- [x] Always use curly braces `{}` for a single-line block.
- [x] All test functions need to cover 1D arrays.

ThreeMonth03 committed:
Feat: Implement the basic statistic operation

ThreeMonth03 commented on cpp/modmesh/buffer/SimpleArray.hpp:
```diff
@@ -135,6 +147,102 @@ class SimpleArrayMixinCalculators
+        {
+            return acopy.select_kth(n / 2);
+        }
+        auto v1 = acopy.select_kth(n / 2 - 1);
+        auto v2 = acopy.select_kth(n / 2);
```
Implement the quick selection algorithm in `small_vector`.

ThreeMonth03 commented on tests/test_math.py:
```diff
@@ -159,11 +159,11 @@ def test_operator_comparison(self):
-
-        self.assertEqual(cplx1 < cplx2, norm1 < norm2)
-        self.assertEqual(cplx1 > cplx2, norm1 > norm2)
+        def compare():
+            if self.real1 == self.real2:
+                return self.imag1 <= self.imag2
+            return self.real1 <= self.real2
+        self.assertEqual(cplx1 < cplx2, compare())
```
Modify the operator < with  lexicographical order.

yungyuc replied:
Leave a code comment here in the test code for the lexicographical ordering of the real and imaginary parts of the complex number.

ThreeMonth03 commented on tests/test_buffer.py:
```diff
@@ -998,6 +998,176 @@ def test_abs(self):
+        sarr = modmesh.SimpleArrayFloat64(array=nparr)
+        npstd = np.var(nparr)
+        sstd = sarr.var()
+        self.assertEqual(npstd, sstd)
```
Add the testcase about slicing array.

yungyuc replied:
Good fix.  But it should be moved to a distinct PR.

ThreeMonth03 commented on tests/test_buffer.py:
```diff
@@ -998,6 +998,176 @@ def test_abs(self):
+        sstd = sarr.std()
+        self.assertEqual(npstd, sstd)
+
+        nparr = np.arange(24, dtype='complex128').reshape((2, 3, 4))
+        npimg = nparr.copy()
+        np.random.shuffle(nparr)
+        np.random.shuffle(npimg)
+        nparr = npimg + 1j * npimg
+        sarr = modmesh.SimpleArrayComplex128(array=nparr)
+        npstd = np.std(nparr)
+        sstd = sarr.std()
+        self.assertEqual(npstd, sstd)
```
Add the testcase of complex array.

ThreeMonth03 commented on cpp/modmesh/math/Complex.hpp:
```diff
@@ -44,6 +44,20 @@ struct ComplexImpl
     T real_v;
     T imag_v;
 
+    ComplexImpl() = default;
+
+    ComplexImpl(T r, T i)
+        : real_v(r)
+        , imag_v(i)
+    {
+    }
+
+    ComplexImpl(T t)
+        : real_v(t)
+        , imag_v(0.0)
+    {
+    }
```
Allow the static casting from integer to complex.

ThreeMonth03 commented on cpp/modmesh/buffer/small_vector.hpp:
```diff
@@ -270,6 +270,66 @@ class small_vector
+                right = store;
+            }
+        }
+    }
```
Quick selection algorithm.

yungyuc replied:
I did not look into the implementation, but it is great to have it here.  Let's do it this way for now and improve in the future.

ThreeMonth03 commented on cpp/modmesh/buffer/pymod/wrap_SimpleArray.cpp:
```diff
@@ -87,10 +88,32 @@ class MODMESH_PYTHON_WRAPPER_VISIBILITY WrapSimpleArray
+                            ConcreteBuffer::construct(
+                                view_nbytes,
+                                view_ptr,
+                                std::move(remover));
```
Fix the initializer from numpy to simplearray.
If numpy array is sliced, `arr_in.nbytes()` returns the nbytes after sliced. It doesn't return the nbytes of the buffer.

ThreeMonth03 replied:
Currently, `view_nbytes` is the nbytes from `view_ptr` to the end of the buffer, and `view_ptr` is `buffer.begin() + offset`.

ThreeMonth03 commented on cpp/modmesh/buffer/SimpleArray.hpp:
```diff
@@ -611,17 +719,6 @@ class SimpleArray
-                throw std::runtime_error(Formatter() << "SimpleArray: shape byte count " << nbytes
-                                                     << " differs from buffer " << buffer->nbytes());
-            }
-
```
This check will fail if we modify the nbytes as the size of the buffer.

ThreeMonth03 commented on cpp/modmesh/buffer/SimpleArray.hpp:
```diff
@@ -135,6 +147,102 @@ class SimpleArrayMixinCalculators
+    real_type var(size_t ddof) const
+    {
+        auto athis = static_cast<A const *>(this);
+        size_t n = 1;
+        for (size_t i = 0; i < athis->ndim(); ++i)
+        {
+            n *= athis->shape(i);
+        }
```
This is the size calculated by shape. 
I think we should open additional discussion and pull request to reimplement the metadata `size()`, `nbytes()`, `buffer_size()` and `buffer_nbytes()`. Currently, the modification of this pull request may cause the inconsistence between `SimpleArray::nbytes()` and `Numpy::nbytes()`.

ThreeMonth03 commented on cpp/modmesh/buffer/SimpleArray.hpp:
```diff
@@ -135,6 +147,102 @@ class SimpleArrayMixinCalculators
+        do {
+            sum += athis->at(sidx);
+            ++total;
+        } while (athis->next_sidx(sidx));
```
Consider the stride of different dimensions.

ThreeMonth03 reviewed (commented):
@yungyuc You could review this pull request if the CI pass. Thank you.

yungyuc commented on cpp/modmesh/buffer/SimpleArray.hpp:
```diff
@@ -135,6 +147,102 @@ class SimpleArrayMixinCalculators
+        small_vector<T> acopy(n);
+        auto sidx = athis->first_sidx();
+        size_t i = 0;
+        do {
```
Curly braces `{` and `}` should use a standalone lines.

yungyuc commented on cpp/modmesh/buffer/small_vector.hpp:
```diff
@@ -270,6 +270,66 @@ class small_vector
+        iterator mid = left + (right - left) / 2;
+        iterator last = right - 1;
+        if (*mid < *first)
+            std::iter_swap(mid, first);
```
Always use curly braces `{}` for blocks.

ThreeMonth03 replied:
Fix.

yungyuc commented on cpp/modmesh/buffer/small_vector.hpp:
```diff
@@ -270,6 +270,66 @@ class small_vector
+        return mid;
+    }
+
+    iterator quick_select(iterator first, iterator last, size_t k)
```
We need C++ unit tests for this.  Please create an issue to add the C++ unit tests for `small_vector::quick_select()`.

yungyuc commented on cpp/modmesh/math/Complex.hpp:
```diff
@@ -121,13 +135,23 @@ struct ComplexImpl
 bool operator<(const ComplexImpl<T> & lhs, const ComplexImpl<T> & rhs)
 {
-    return lhs.norm() < rhs.norm();
+    // Reference1: https://github.com/numpy/numpy/issues/12943
```
It is good to have related reference like these, but they do not specify how numpy compares complex number (directly, not by using absolute number).

In the comment of the number issue 12943, https://github.com/numpy/numpy/issues/12943#issuecomment-969859179 , the use of lexicographic ordering of the real and imaginary parts was mentioned, it will be better to find a more explicit reference of the specification and also write down the code comments here.

yungyuc commented on cpp/modmesh/buffer/pymod/wrap_SimpleArray.cpp:
```diff
@@ -87,10 +88,37 @@ class MODMESH_PYTHON_WRAPPER_VISIBILITY WrapSimpleArray
+
+                        char * base_ptr = static_cast<char *>(owner.mutable_data());
+                        char * view_ptr = static_cast<char *>(arr_in.mutable_data());
+                        const ptrdiff_t offset_bytes = view_ptr - base_ptr;
```
It seems that you piggybacked a fix to issue #527 in this PR without saying it explicitly.  Thanks for the fix, but it is not good to mix two things in one PR.  Please separate this fix (to issue #527, along with the related unit test code) in a different PR.

yungyuc commented on cpp/modmesh/buffer/SimpleArray.hpp:
```diff
@@ -135,6 +147,102 @@ class SimpleArrayMixinCalculators
 public:
 
     using value_type = typename internal_types::value_type;
+    using real_type = typename detail::select_real_t<value_type>::type;
```
Maybe `select_real` is more concise.  But `select_real_t` is fine.  You decide.

yungyuc reviewed (changes requested):
Thank you, @ThreeMonth03 .  Points to address:

- [x] Format: Curly braces `{` and `}` should use standalone lines.
- [x] Format: Always use curly braces `{}` for blocks.
- [x] Code comment: Clarify the reference of the numpy use of lexicographic ordering of the real and imaginary parts and document in modmesh code.  Both in C++ implementation and Python unit tests.
- [x] Separate the fix to issue #527, along with the related unit test code, in a different PR.

Task after merging:

- [ ] Create an issue to add the C++ unit tests for `small_vector::quick_select()`.

@Gene0315 we are still waiting for your document for the complex number variance formula.

ThreeMonth03 commented on cpp/modmesh/math/Complex.hpp:
```diff
@@ -121,13 +135,24 @@ struct ComplexImpl
-    return lhs.norm() < rhs.norm();
+    // Reference1: https://numpy.org/devdocs/reference/generated/numpy.sort.html
+    // Reference2: https://github.com/numpy/numpy/issues/12943
+    // Reference3: https://stackoverflow.com/questions/52481376/numpy-minimum-for-complex-numbers
```
Add more comments.

yungyuc replied:
Please add clear description about what it is doing here along with the reference.  The `numpy.sort()` document is a good example.

> The sort order for complex numbers is lexicographic.

In the context of this function, you may write something like:

```cpp
// Use lexicographical order for comparing complex number, so that it matches the implementation of numpy.
// See https://numpy.org/devdocs/reference/generated/numpy.sort.html and other discussions:
// 1. ...
// 2. ...
```

ThreeMonth03 commented on cpp/modmesh/buffer/SimpleArray.hpp:
```diff
@@ -135,6 +147,106 @@ class SimpleArrayMixinCalculators
+        value_type sum = 0;
+        auto sidx = athis->first_sidx();
+        int64_t total = 0;
+        do
+        {
+            sum += athis->at(sidx);
```
Make the standalone curly braces.

ThreeMonth03 commented on tests/test_math.py:
```diff
@@ -158,12 +158,19 @@ def test_operator_div(self):
+https://github.com/numpy/numpy/issues/12943
+Reference3:
+https://stackoverflow.com/questions/52481376/numpy-minimum-for-complex-numbers
+        '''
```
Add the comments.

ThreeMonth03 reviewed (commented):
@yungyuc You can review this pull request when you are available. Thank you.

yungyuc commented on cpp/modmesh/buffer/SimpleArray.hpp:
```diff
@@ -135,6 +147,106 @@ class SimpleArrayMixinCalculators
+        {
+            sum += athis->at(sidx);
+            ++total;
+        } while (athis->next_sidx(sidx));
```
You missed the closing curly brace `}` here.

ThreeMonth03 replied:
I've tried to modify the closing curly brace. However, it cannot pass the linter when the closing curly brace is standalone.
https://github.com/solvcon/modmesh/blob/4e5991bd00b91a2e398494d158c4a0e88c82949e/cpp/modmesh/toggle/pymod/wrap_Toggle.cpp#L422

ThreeMonth03 replied:
For example, if I modify the closing curly brace to the following pattern:
```c++
        do
        {
            acopy[i] = athis->at(sidx);
            ++i;
        }
        while (athis->next_sidx(sidx));
```

Here is the output of the `make lint`:
```console
zhangjunshideMacBook-Air:modmesh abao$ make lint
...
clang-format --dry-run cpp/modmesh/buffer/ConcreteBuffer.hpp:
clang-format --dry-run cpp/modmesh/buffer/SimpleArray.cpp:
clang-format --dry-run cpp/modmesh/buffer/SimpleArray.hpp:
cpp/modmesh/buffer/SimpleArray.hpp:167:10: warning: code should be clang-formatted [-Wclang-format-violations]
        }
         ^
clang-format --dry-run cpp/modmesh/buffer/SimpleCollector.hpp:
clang-format --dry-run cpp/modmesh/buffer/buffer.hpp:
clang-format --dry-run cpp/modmesh/buffer/pymod/SimpleArrayCaster.hpp:
clang-format --dry-run cpp/modmesh/buffer/pymod/TypeBroadcast.hpp:
...
```

yungyuc replied:
Ok, we comply with clang-format.  Thank you for clarifying this.

yungyuc reviewed (changes requested):
Remaining defects:

- [x] Format: missed closing curly brace `}`.
- [x] Comment for lexicographical ordering.

ThreeMonth03 commented on tests/test_math.py:
```diff
@@ -158,12 +158,21 @@ def test_operator_div(self):
+1.
+https://github.com/numpy/numpy/issues/12943
+2.
+https://stackoverflow.com/questions/52481376/numpy-minimum-for-complex-numbers
```
Add the reason and reference to explain the modification of the compare operation.

yungyuc replied:
Please use comment mark `#`.  The docstring format does not look nice here.

ThreeMonth03 reviewed (commented):
@yungyuc You can review this pull request when you are available. Thank you.

yungyuc commented on cpp/modmesh/math/Complex.hpp:
```diff
@@ -121,13 +135,31 @@ struct ComplexImpl
 bool operator<(const ComplexImpl<T> & lhs, const ComplexImpl<T> & rhs)
 {
-    return lhs.norm() < rhs.norm();
+    /*
```
This comment block is clear.

yungyuc reviewed (changes requested):
- [x] Format: use comment mark instead of docstrings for the long comments in `test_math.py` complex number tests.

ThreeMonth03 commented on tests/test_math.py:
```diff
@@ -159,11 +159,21 @@ def test_operator_comparison(self):
+        # 1.
+        # https://github.com/numpy/numpy/issues/12943
+        # 2.
+        # https://stackoverflow.com/questions/52481376
```
Fixed.

ThreeMonth03 reviewed (commented):
@yungyuc I've modified the python comment with `#`. Thank's for your suggestion.

yungyuc reviewed (approved):
Looks good

yungyuc committed:
Typeset comments for the lexicographic ordering

In C++, use the doxygen format for the lexicographic comments for the comparison operators.  In Python (test), make the comments read more fluent.

yungyuc commented on cpp/modmesh/math/Complex.hpp:
```diff
@@ -117,17 +131,34 @@ struct ComplexImpl
 }; /* end struct ComplexImpl */
 
-// These comparison operator would be used in SimpleArray::min(), SimpleArray::max().
+/**
```
Edit and typeset the comment block. Use doxygen format.

yungyuc commented on tests/test_math.py:
```diff
@@ -159,11 +159,18 @@ def test_operator_comparison(self):
-
-        self.assertEqual(cplx1 < cplx2, norm1 < norm2)
-        self.assertEqual(cplx1 > cplx2, norm1 > norm2)
+        # Numpy uses lexicographic ordering to compare complex numbers, as
```
Edit the comment block for more fluent description.

yungyuc reviewed (commented):
I edited and typeset the lexicographic ordering comments before merging.
